### PR TITLE
[12.0][ADD] helpdesk_mgmt_timesheet

### PR DESCRIPTION
**Pull request by dalonsod**
This module adds timesheet support for helpdesk tickets. 

Supersedes #121 . Original commits were squashed for cleaning purposes in a single one. Since original PR, these changes have been made:
- rebased in order to retrieve `helpdesk_mgmt_project` dependency.
- changed manifest dependencies and `account.analytic.account` references have become `project.project` ones.
- some file renaming.
- added full ticket support (menu, views, permissions, report) to `account.analytic.line` model.
- improved `search` and `tree` views for tickets.
- added demo data.

It seems a lot of changes, but original work is still present :-)

### CODE REVIEW

**Comment by dalonsod** [comment]
@oscars8a @EDallas89 @marcelsavegnago could you review? Thanks!

@DarioLodeiros runbot shows an error status, but it's actually working and I'm not able to find out any error message. Could you check?

To everyone, I've slightly changed ticket requirements to the "allow timesheet" functionality. Since `helpdesk_mgmt_project` already adds `project_id` to a to ticket, and that's actually the main requirement to timesheet management, if present in a ticket automatically timesheets are allowed. Then, I redefined "allow timesheet" for helpdesk teams: now is used to make the project required in a ticket. What dou you think?

**Comment by DarioLodeiros** [comment]
I can't see the problem, @pedrobaeza can you review this runbot fail? and coverage?

File: helpdesk_mgmt_timesheet/views/helpdesk_team_view.xml [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+        <field name="model">helpdesk.ticket.team</field>
+        <field name="inherit_id" ref="helpdesk_mgmt.view_helpdesk_team_form" />
+        <field name="priority" eval="20" />
+        <field name="arch" type="xml">
+            <xpath expr="//field[@name='alias_user_id']" position="after">
+                <field name="allow_timesheet" type="checkbox" />
+                <field name="default_project_id" attrs="{'invisible':[('allow_timesheet', '=', False)]}" help=" Change the Default Project will not have retroactive effects."/>
+                <button name="action_clear" help=" Clean Default Project" type="object" icon="fa-trash" attrs="{'invisible':['|',('allow_timesheet', '=', False),('default_project_id', '=', False)]}" class="oe_highlight">
+                    Clean Default Project
+                </button>
```
Comment (hbrunn) [inline helpdesk_mgmt_timesheet/views/helpdesk_team_view.xml]: why do we need this button?

File: helpdesk_mgmt_timesheet/views/helpdesk_ticket_view.xml [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+        <field name="name">timesheet.helpdesk.ticket.form.view</field>
+        <field name="model">helpdesk.ticket</field>
+        <field name="inherit_id" ref="helpdesk_mgmt.ticket_view_form" />
+        <field name="priority" eval="20" />
+        <field name="arch" type="xml">
+            <xpath expr="//field[@name='partner_email']" position="after">
+                <field name="allow_timesheet" invisible="1" />
+            </xpath>
+            <xpath expr="//page[@name='description']" position="after">
+                <page string="Timesheets" attrs="{'invisible':[('project_id', '=', False)]}">
```
Comment (hbrunn) [inline helpdesk_mgmt_timesheet/views/helpdesk_ticket_view.xml]: this should be restricted to timesheet users I think

File: helpdesk_mgmt_timesheet/models/hr_timesheet.py [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+
+class AccountAnalyticLine(models.Model):
+    _inherit = 'account.analytic.line'
+
+    ticket_id = fields.Many2one(
+        comodel_name='helpdesk.ticket',
+        string='Ticket',
+        groups="helpdesk_mgmt.group_helpdesk_user",
+    )
```
Comment (hbrunn) [inline helpdesk_mgmt_timesheet/models/hr_timesheet.py:15]: I suggest to add a constraint enforcing that this belongs to the project

File: helpdesk_mgmt_timesheet/security/helpdesk_mgmt_timesheet_security.xml [hunk 4]
```diff
@@ -0,0 +1,10 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo>
+
+    <data noupdate="0">
+        <record id="helpdesk_mgmt.group_helpdesk_user_own" model="res.groups">
+            <field name="implied_ids" eval="[(4, ref('hr_timesheet.group_hr_timesheet_user'))]"/>
```
Comment (hbrunn) [inline helpdesk_mgmt_timesheet/security/helpdesk_mgmt_timesheet_security.xml]: I don't think we should have that. When we want a user to be able to record time on tickets, they should have both groups assigned manually

File: helpdesk_mgmt_timesheet/views/helpdesk_ticket_view.xml [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+    <record id="helpdesk_ticket_view_search" model="ir.ui.view">
+        <field name="name">helpdesk.ticket.view.search</field>
+        <field name="model">helpdesk.ticket</field>
+        <field name="inherit_id" ref="helpdesk_mgmt.helpdesk_ticket_view_search" />
+        <field name="arch" type="xml">
+            <filter name="last_week" position="before">
+                <filter name="timesheet_allowed" string="Timesheet allowed"
+                        domain="[('project_id', '!=', False)]"/>
+                <separator/>
+                <filter name="overtimed" string="Overtimed"
```
Comment (hbrunn) [inline helpdesk_mgmt_timesheet/views/helpdesk_ticket_view.xml]: I don't know this word. How about 'late'?

**Review by marcelsavegnago — COMMENTED** [review]
The task is being created without linking to the product. I believe it is important to take on the project that is defined in the ticket.

**Review by hbrunn — CHANGES_REQUESTED** [review]
runbot is red because of


```
2020-07-30 04:07:02,187 161 WARNING openerp_test odoo.models: project.project.create() with unknown fields: project_id
```

in the all logs. Coverage only runs when both travis and runbot succeed

File: helpdesk_mgmt_timesheet/views/helpdesk_team_view.xml (hunk 1 above)
Comment (DarioLodeiros) [reply helpdesk_mgmt_timesheet/views/helpdesk_team_view.xml]: +1

Comment (dalonsod) [reply helpdesk_mgmt_timesheet/views/helpdesk_team_view.xml]: OK, I'll review it, and remove if does not make sense. This development started with some initial functionalities, I'd preserved some of them without knowing their purpose or intention.

File: helpdesk_mgmt_timesheet/views/helpdesk_ticket_view.xml (hunk 2 above)
Comment (dalonsod) [reply helpdesk_mgmt_timesheet/views/helpdesk_ticket_view.xml]: I agree, specially if helpdesk and timesheet user security groups are not linked any more

File: helpdesk_mgmt_timesheet/models/hr_timesheet.py (hunk 3 above)
Comment (dalonsod) [reply helpdesk_mgmt_timesheet/models/hr_timesheet.py:15]: I think I put it the view, I'll change it to the field definition if possible

File: helpdesk_mgmt_timesheet/security/helpdesk_mgmt_timesheet_security.xml (hunk 4 above)
Comment (dalonsod) [reply helpdesk_mgmt_timesheet/security/helpdesk_mgmt_timesheet_security.xml]: Yes, I had a big doubt with this, removing this inheritance, we can maintain both ticket users with and without timesheet access. I'll remove it and review every timesheet access within this module.

File: helpdesk_mgmt_timesheet/views/helpdesk_ticket_view.xml (hunk 5 above)
Comment (dalonsod) [reply helpdesk_mgmt_timesheet/views/helpdesk_ticket_view.xml]: Sorry for my english :( I was looking for "this ticket has extra time" in a single word

**Comment by dalonsod** [comment]
> 
> 
> runbot is red because of
> 
> ```
> 2020-07-30 04:07:02,187 161 WARNING openerp_test odoo.models: project.project.create() with unknown fields: project_id
> ```
> 
> in the all logs. Coverage only runs when both travis and runbot succeed

I'll look for this reference on tests, thanks!

**Comment by dalonsod** [comment]
> 
> 
> The task is being created without linking to the product. I believe it is important to take on the project that is defined in the ticket.

For _product_, do you really want to say "project"? Anyway, project and task management don't belong to this module. Maybe you've detected a problem related to `helpdesk_mgmt_project` module, if I'm right.

File: helpdesk_mgmt_timesheet/views/helpdesk_team_view.xml (hunk 1 above)
Comment (dalonsod) [reply helpdesk_mgmt_timesheet/views/helpdesk_team_view.xml]: Button removed

File: helpdesk_mgmt_timesheet/models/hr_timesheet.py (hunk 3 above)
Comment (dalonsod) [reply helpdesk_mgmt_timesheet/models/hr_timesheet.py:15]: Done!

**Comment by dalonsod** [comment]
@hbrunn comments attended

**Comment by marcelsavegnago** [comment]
> For _product_, do you really want to say "project"? Anyway, project and task management don't belong to this module. Maybe you've detected a problem related to `helpdesk_mgmt_project` module, if I'm right

@dalonsod You're right, I got confused with scope. In fact it would be something in the helpdesk_mgmt_project.

Yes, i really want say project..

**Comment by dalonsod** [comment]
> @dalonsod You're right, I got confused with scope. In fact it would be something in the helpdesk_mgmt_project.
> 
> Yes, i really want say project..

Exactly. Open #152

**Comment by dalonsod** [comment]
@DarioLodeiros I've found an error when accessing timesheets for a user without permissions in Helpdesk, due to `search` view additions (`filter` nodes referring `ticket_id`) to `account.analytic.line`. There are two different ways to solve this: adding `groups="helpdesk_mgmt.group_helpdesk_user"` to every filter definition within the view, or forcing access rights for the whole view to the group above. What is the best approach, is there any OCA convention for this? Form and tree views work correctly without any change.

**Comment by dalonsod** [comment]
Any news on this? If approved, I could also migrate it to v13. Thanks!

**Review by marcelsavegnago — APPROVED** [review]
Funcional review Ok

**Review by AntoniRomera — APPROVED** [review]

**Review by DarioLodeiros — APPROVED** [review]
Functional y Technical Review

**Review by flachica — APPROVED** [review]

**Comment by flachica** [comment]
> 13.

I migrate to 13.0. I need it

**Comment by dalonsod** [comment]
@DarioLodeiros this addon is still unavailable to be translated: https://translation.odoo-community.org/projects/helpdesk-12-0/ . However, in v13 it's OK. Could you check?

**Comment by dalonsod** [comment]
> 
> 
> @DarioLodeiros this addon is still unavailable to be translated: https://translation.odoo-community.org/projects/helpdesk-12-0/ . However, in v13 it's OK. Could you check?

@DarioLodeiros any news on this?

**Pull request merged**